README.md
# Markdown Planner

A simple approach to planning ahead.

## Features

* Easily write plans and mark things to do in Markdown files
* See at a glance what is to be done next in the **TO DO** view
* Organize tasks by urgency and importance
* Specify due dates and recurring tasks
* Synchronize with Google Tasks and Google Calendar

## Planning In Markdown?

Planning is often done using specialized tools ranging from basic todo list managers like Google Tasks to complex project management tools.

For most planning, however, slightly enhanced text files have many advantages.

Text files are easy to write, to read, to search, to print and to share with others. They are maintainable and preservable, with no obsolescent proprietary formats or software to worry about.

Markdown is a well supported standard for text processing, particularly in Visual Studio Code, and provides a good platform to build upon.

<!--
By writing our plans as Markdown files we get a lot of functionality for free, for example:

* Outline (look in the Explorer area on your left side) that makes it easy to understand and navigate your plan 
* Print preview (open with `ctrl-K V`)
* Useful functions to quickly enter or modify lists, headers and so on

.. and much more.


## Planning: What To Do Next?

The main aim of a planning system is to select the next action to perform in order to achieve a certain goal.

To do so, we usually do not need to evalute precisely the value, cost and dependencies of every possible action, we just need to provide enough information to calculate a rough ordering of priorities.
-->

## Example: A Jogging Plan

So, how does a plan look like in Markdown? 

For example, a plan to become a runner might be stated as follows:

```markdown
Become a Long Distance Runner SOON IMPORTANT:

* Choose a training program
* Buy running shoes TODO
* Training
    * Short Run EVERY WEDNESDAY EVERY FRIDAY
    * Long Run EVERY SUNDAY
* Events
    * Florence Half-Marathon DUE 2018-04-15
    * Florence Marathon DUE 2018-11-25
```

The goal (Become a Long Distance Runner) is stated first and marked as important and to be started soon.

Then follows the list of tasks required to accomplish the goal, ordered by their priority (first choose a training program then buy suitable shoes).

The plan also specifies recurring tasks (weekly running days) and some specific events (run the half or full marathon).

## Planning on a Small Scale

To prioritize the steps required to achieve a simple goal, just order them in a list, with the next one to be done at the top:

* My lofty goal, in order to achieve it:
    * Do this first
    * Then this
    * and finally this
        * doing this first
        * and then this

### Rearranging Tasks

Keep rearranging the task till it feels right.

To move a line or a selected block of text up and down use: `Alt + ↑` and `Alt + ↓`.

To select a block of text use `ctrl-a` (`ctrl` becomes `cmd` on Mac). It will select the smallest logical block of text that surrounds the cursor. Repeating the command will select progressively larger blocks of texts. Use `ESC` to deselect. 

To move around the selected block then either cut (`ctrl-X`) and paste (`ctrl-V`) it or move it up and down using `Alt + ↑` and `Alt + ↓`.

## Planning on a Large Scale
 
To handle many different and/or unrelated goals and actions, it is useful to organize them by two simple dimensions: urgency and importance.

This gives rise to the famous [Eisenhower Matrix](https://en.wikipedia.org/wiki/Time_management#The_Eisenhower_Method):


|                | Important | Not Important |
| -------------- | --------- | ------------- |
| **Urgent**     | 1         | 2             |
| **Not Urgent** | 3         | 4             |

The matrix split tasks in four groups of decreasing priority:

1. Urgent and Important
    * Tasks that should be attended to immediately
2. Urgent and Non Important
    * Tasks that should be delegated. If that is not possible, they are second in line to be dealt with.
3. Not Urgent and Important
    * Long term goals on which to focus once the urgent stuff has been dealt with 
4. Not Urgent and Not Important
    * The clutter that should be ignored

### Marking Tasks By Importance

To mark a task as important, write **IMPORTANT** or **BIG** (but not **BIGLY**):

    Become a Better Human Being IMPORTANT

### Marking Tasks By Urgency

To mark a task as being urgent write **NOW** (to be done immediately), **SOON** (to be done in the next 3 days) or **TODO** (to be done in the next 11 days):

    Learn all about markdown SOON

If there is a specific date by which a task is due, specify a **DUE** date. To quickly enter a DUE date, type `due` and then `ctrl-space` and select the `due` code snippet:

    Pay phone bill DUE 2018-03-30

Note that dates are specified in international format: **YYYY-MM-DD**.

If the tasks is recurrent, you can specify on which day of the month or the week it takes place. 

To quickly enter a weekly task: type the first three letters of the weekday (`sun`, `mon`, `tue` ..), then `ctrl-space` to select the corresponding code snippet:

    Go shopping EVERY MONDAY EVERY THURSDAY
    Check online bank account EVERY 1

Note that you can have multiple markers per task.

## What To Do Next?

All urgent or important tasks in the current file are displayed in the **TO DO** view in the Explorer area (on the left of the screen).

Click on the task to jump to its position in the document.

The tasks are ordered by their priority class and inside every class by their urgency.

All tasks that are due in less than 11 days are considered urgent.

Tasks that neither urgent nor important are not displayed.

## Import and Export Planning Information

One problem with planning is that the required information is often scattered around.

Some of it will be in your online or offline calendar, a lot will be in your emails.

With most WWW based email programs (like GMail) you can reference an email from Markdown by just cutting and pasting its URL.

This extension provides a couple of commands that help to integrate external information with your Markdown Planner files:

<!--
One A convenient daily workflow (for Google users at least) to stay one step ahead of your commitments is to go through your email inbox, every evening, and clean it completely:

* Delete all the junk emails
* Answer all the emails that can be answered quickly
* Transform all emails that you have to act upon in tasks (in gmail, look under **More** for **Add to Tasks**) and then archive them
* Archive all emails that you do not need to act upon but that contains useful information you need to keep
* Rejoice in the emptiness of your email box :-)
* Then use Then convert 

To integrate all this information in your markdown planner files, we need a way of moving it back and
-->


### Save a Task on Google Calendar

`ctrl-e + ctrl c` (Mnemonic: **e**xport to **c**alendar) to save the task under the cursor in Google Calendar.

### Retrieve Google Tasks

`ctrl-i + ctrl t` (Mnemonic: **i**mport from google **t**asks) to import your task lists from Google Tasks.

The command will simply open the web page https://quid2.org/tasks in your browser.

After logging into your Google account, your tasks lists will be converted to Markdown Planner format. 

Just cut-and-paste them from the web page to your local markdown planner file.

Note that your task lists are converted and not deleted. 

## Reference

All commands provided by this extensions are under the **Planner** prefix (to find them `ctrl-p` and then enter `> Planner`).

| Command                      | ShortCut          |
| ---------------------------- | ----------------- |
| Select Block                 | `ctrl-a`          |
| Save Task to Google Calendar | `ctrl-e + ctrl c` |
| Show Task Date in Google Calendar | `ctrl-s + ctrl c` |
| Retrieve Google Tasks        | `ctrl-i + ctrl t` |
| Open Markdow Preview         | `ctrl-k v`        |
| Move Selected Block Up       | `Alt + ↑`         |
| Move Selected Block Down     | `Alt + ↓`         |

    

| Snippet         | ShortCut                      |
| --------------- | ----------------------------- |
| DUE today       | `due`                         |
| EVERY MONDAY .. | `mon tue wed thu fri sat sun` |

## Known Issues

See [Issues](https://github.com/tittoassini/vscode-planner/issues).

## Release Notes

See the [Change Log](https://github.com/tittoassini/vscode-planner/blob/master/CHANGELOG.md).

## Acknowledgements

<div>Extension Icon made by <a href="http://www.freepik.com" title="Freepik">Freepik</a> from <a href="https://www.flaticon.com/" title="Flaticon">www.flaticon.com</a> is licensed by <a href="http://creativecommons.org/licenses/by/3.0/" title="Creative Commons BY 3.0" target="_blank">CC 3.0 BY</a>.</div>



CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

## [0.3.7](https://github.com/tittoassini/vscode-planner/compare/v0.3.6...v0.3.7) (2020-01-30)
- Google Tasks are now imported from https://quid2.org (HTTPS) rather than http://quid2.org

<a name="0.3.3"></a>
## [0.3.3](https://github.com/tittoassini/vscode-planner/compare/v0.3.2...v0.3.3) (2019-03-11)



<a name="0.3.2"></a>
## [0.3.2](https://github.com/tittoassini/vscode-planner/compare/v0.3.1...v0.3.2) (2019-03-11)
- Keywords (TODO, SOON ..) are now displayed as keywords
- Added 'Show Task Date in Google Calendar'
- Added TODO keyword
- Fixed bug: TO DO view not displayed for first markdown file displayed
- Update package dependencies

<a name="0.3.1"></a>
## 0.3.1 (2018-04-01)
- Improved block selection
 
## 0.2 (2018-03-31)
- Initial release

src/priority.doctest.ts
const millisecsPerDay = 24 * 60 * 60 * 1000;

/**
 *  daysLeft("no rush") // => 3650
 *  daysLeft("TODO") // => 11
 *  daysLeft("SOON") // => 3
 *  daysLeft("NOW") // => 0
 *  daysLeft("SOON NOW SOON") // => 0
 *  daysLeft("DUE 2018-1-11",new Date(2018,0,1))  // => 10
 *  daysLeft("DUE 2018-1-11 NOW",new Date(2018,0,1))  // => 0
 *  daysLeft("DUE 2018-1-11 SOON",new Date(2018,0,1))  // => 3
 *  daysLeft("DUE 2018-1-22 DUE 2018-1-11",new Date(2018,0,1))  // => 10
 *  daysLeft("DUE 2018-0-32",new Date(2018,0,1))  // => 3650
 *  daysLeft("EVERY 11 EVERY 9",new Date(2018,1,11))  // => 0
 *  daysLeft("EVERY 11 EVERY 3",new Date(2018,1,1))  // => 2
 *  daysLeft("EVERY 3",new Date(2017,11,28))  // => 6
 *  daysLeft("EVERY 3",new Date(2017,1,27))  // => 4
 *  daysLeft("EVERY 33",new Date(2018,0,1))  // => 3650
 *  daysLeft("EVERY FRIDAY",new Date(2018,2,30))  // => 0
 *  daysLeft("EVERY MONDAY",new Date(2018,2,30))  // => 3
 *  daysLeft("EVERY WEDNESDAY",new Date(2018,2,30))  // => 5
 * 
 * @return the number of days left to perform task (it can be negative)
 */
function daysLeft(line: string, since = new Date()): number {

    function add(dt: Date) { days.push((dt.getTime() - since.getTime()) / millisecsPerDay); }

    function addMatches(regex: RegExp, f: (r: RegExpExecArray) => Date | undefined) {
        regex.lastIndex = 0; // BAD, MODIFYING A GLOBAL VALUE
        var due;
        while ((due = regex.exec(line)) !== null) {
            const dt = f(due);
            if (dt) {
                add(dt);
            }
        }
    }

    var days: number[] = [];

    if (nowRegex.test(line)) { add(since); }

    if (soonRegex.test(line)) { days.push(3); }

    if (todoRegex.test(line)) { days.push(11); }

    addMatches(dueRegex, function (r) {
        const year = Number(r[1]);
        const month = Number(r[2]);
        const day = Number(r[3]);
        if (month > 12 || day > 31) { return; }
        return new Date(year, month - 1, day);
    });

    addMatches(everyDayRegex, function (r) {
        const day = Number(r[1]);
        if (day > 31) { return; }
        const date = new Date(since.getTime());
        date.setDate(day);
        if (date < since) {
            date.setMonth(date.getMonth() + 1);
        }
        return date;
    });

    addMatches(everyDayOfWeekRegex, function (r) {
        const day = ["SUNDAY", "MONDAY", "TUESDAY", "WEDNESDAY", "THURSDAY", "FRIDAY", "SATURDAY"].indexOf(r[1]);
        const date = new Date(since.getTime());
        const diff = day - date.getDay();
        const toAdd = diff >= 0 ? diff : diff + 7;
        date.setDate(date.getDate() + toAdd);
        return date;
    });

    if (days.length === 0) { return 3650; }

    return days.sort(function (a, b) { return a - b; })[0];
}

/**
 *  dueOn("no rush",new Date(2018,0,1)) // => new Date(2027,11,30)
 *  dueOn("NOW",new Date(2018,0,1)) // => new Date(2018,0,1)
 *  dueOn("DUE 2018-1-22 DUE 2018-1-11",new Date(2018,0,1))  // => new Date(2018,0,11)
 * 
 * @param line the line of text corresponding to the task
 * @return the date by which the current line/task is due
*/
export function dueOn(task: string, since = new Date()): Date {
    return new Date(daysLeft(task, since) * millisecsPerDay + since.getTime());
}

function pad2(s: string): string { return ("0" + s).slice(-2); }

/**
 * compactDate(new Date(2018,3,11)) // => "20180411"
 * 
 * @param d the Date to convert 
 * @return the Date in compact international format 
 * */
export function compactDate(d: Date): string { return [d.getUTCFullYear(), pad2("" + (d.getMonth() + 1)), pad2(d.getDate() + "")].join(""); }

/**
 * gcalendarDate(new Date(2018,3,11)) // => "2018/4/11"
 * 
 * @param d the Date to convert
 * @return the Date in the format required for google Calendar URLs
 * */
export function gcalendarDate(d: Date): string { return [d.getUTCFullYear(), d.getMonth() + 1, d.getDate()].join("/"); }

/** 
 * isImportant("no rush") // => false
 * isImportant("this is BIG") // => true
 * isImportant("IMPORTANT") // => true
 * isImportant("..IMPORTANT..") // => true
 * isImportant("really BIG..") // => true
 * isImportant("no BIGLY") // => false
 * isImportant("..IMPORTANTISSIMO..") // => false
 * 
 * @param task
 * @return true if the task is important
 */
function isImportant(task: string): boolean { return importanceRegex.test(task); }

/**
 * priority("BIG and NOW") // => { group: 1, daysLeft: 0 }
 * priority("SOON") // => { group: 2, daysLeft: 3 }
 * priority("DUE 2018-03-15").group // => 2
 * priority("IMPORTANT") // => { group: 3, daysLeft: 3650 }
 * priority("no big deal").group // => 4
 * 
 * @param task the task
 * @return the priority level group (1..4) and number of days left before the task is due 
 */
export function priority(task: string): Priority {
    const left = daysLeft(task);
    const urgent = left <= 11;
    const important = isImportant(task);
    const group = urgent ? (important ? 1 : 2) : (important ? 3 : 4);
    return { group: group, daysLeft: left };
}

export interface Priority {
    group: number;      // Eisenhower Matrix Group 1-4 
    daysLeft: number;  // Number of days left before task is due
}

const importanceRegex = /\b(BIG|IMPORTANT)\b/;
export const soonRegex = /\b(SOON)\b/;
export const todoRegex = /\b(TODO)\b/;
export const nowRegex = /\b(NOW)\b/;
export const dueRegex = /DUE\s+(\d{4})-(\d{1,2})-(\d{1,2})/g;
export const everyDayRegex = /EVERY\s+(\d{1,2})/g;
export const everyDayOfWeekRegex = /EVERY\s+(SUNDAY|MONDAY|TUESDAY|WEDNESDAY|THURSDAY|FRIDAY|SATURDAY)/g;

const markers = [importanceRegex, soonRegex, todoRegex, nowRegex, dueRegex, everyDayRegex, everyDayOfWeekRegex];

/** 
 * removeMarkers("Task DUE 2018-1-22 DUE 2018-1-11 SOON NOW") // => "Task"
 * removeMarkers("Task DUE 2018-1-22 DUE 2018-1-11 SOON NOW EVERY SUNDAY to do") // => "Task      to do"
 * 
 * @param line a line of text
 * @return the line without Planner markers and trimmed on both ends
 */
export function removeMarkers(line: string): string {
    return removeRegexs(markers, line);
}

const starters = [/^\#+/, /^(\s|\+|\*)+/];

/**
 * removeStarters("### Header 1") // => "Header 1"
 * removeStarters("Not a header ###") // => "Not a header ###"
 * removeStarters("\t\t  * Indented") // => "Indented"
 *   
 * @param line a line of text
 * @return the line without Markdown header and list markers and trimmed on both ends
 */
export function removeStarters(line: string): string {
    return removeRegexs(starters, line);
}

/** Remove all parts matching the specified regular expressions from text */
export function removeRegexs(regexs: RegExp[], txt: string): string {
    regexs.forEach(rex => txt = txt.replace(rex, ''));
    txt = txt.trim();
    return txt;
}


import * as __test from "tape"

    __test("daysLeft", t => {
      
      t.deepEqual(daysLeft("no rush"), 3650, "3650")
      t.deepEqual(daysLeft("TODO"), 11, "11")
      t.deepEqual(daysLeft("SOON"), 3, "3")
      t.deepEqual(daysLeft("NOW"), 0, "0")
      t.deepEqual(daysLeft("SOON NOW SOON"), 0, "0")
      t.deepEqual(daysLeft("DUE 2018-1-11", new Date(2018, 0, 1)), 10, "10")
      t.deepEqual(daysLeft("DUE 2018-1-11 NOW", new Date(2018, 0, 1)), 0, "0")
      t.deepEqual(daysLeft("DUE 2018-1-11 SOON", new Date(2018, 0, 1)), 3, "3")
      t.deepEqual(daysLeft("DUE 2018-1-22 DUE 2018-1-11", new Date(2018, 0, 1)), 10, "10")
      t.deepEqual(daysLeft("DUE 2018-0-32", new Date(2018, 0, 1)), 3650, "3650")
      t.deepEqual(daysLeft("EVERY 11 EVERY 9", new Date(2018, 1, 11)), 0, "0")
      t.deepEqual(daysLeft("EVERY 11 EVERY 3", new Date(2018, 1, 1)), 2, "2")
      t.deepEqual(daysLeft("EVERY 3", new Date(2017, 11, 28)), 6, "6")
      t.deepEqual(daysLeft("EVERY 3", new Date(2017, 1, 27)), 4, "4")
      t.deepEqual(daysLeft("EVERY 33", new Date(2018, 0, 1)), 3650, "3650")
      t.deepEqual(daysLeft("EVERY FRIDAY", new Date(2018, 2, 30)), 0, "0")
      t.deepEqual(daysLeft("EVERY MONDAY", new Date(2018, 2, 30)), 3, "3")
      t.deepEqual(daysLeft("EVERY WEDNESDAY", new Date(2018, 2, 30)), 5, "5")
      
;t.end()
    })

    __test("dueOn", t => {
      
      t.deepEqual(dueOn("no rush", new Date(2018, 0, 1)), new Date(2027,11,30), "new Date(2027,11,30)")
      t.deepEqual(dueOn("NOW", new Date(2018, 0, 1)), new Date(2018,0,1), "new Date(2018,0,1)")
      t.deepEqual(dueOn("DUE 2018-1-22 DUE 2018-1-11", new Date(2018, 0, 1)), new Date(2018,0,11), "new Date(2018,0,11)")
      
;t.end()
    })

    __test("compactDate", t => {
      
      t.deepEqual(compactDate(new Date(2018, 3, 11)), "20180411", "\"20180411\"")
      
;t.end()
    })

    __test("gcalendarDate", t => {
      
      t.deepEqual(gcalendarDate(new Date(2018, 3, 11)), "2018/4/11", "\"2018/4/11\"")
      
;t.end()
    })

    __test("isImportant", t => {
      
      t.deepEqual(isImportant("no rush"), false, "false")
      t.deepEqual(isImportant("this is BIG"), true, "true")
      t.deepEqual(isImportant("IMPORTANT"), true, "true")
      t.deepEqual(isImportant("..IMPORTANT.."), true, "true")
      t.deepEqual(isImportant("really BIG.."), true, "true")
      t.deepEqual(isImportant("no BIGLY"), false, "false")
      t.deepEqual(isImportant("..IMPORTANTISSIMO.."), false, "false")
      
;t.end()
    })

    __test("priority", t => {
      
      t.deepEqual(priority("BIG and NOW"), { group: 1, daysLeft: 0 }, "{ group: 1, daysLeft: 0 }")
      t.deepEqual(priority("SOON"), { group: 2, daysLeft: 3 }, "{ group: 2, daysLeft: 3 }")
      t.deepEqual(priority("DUE 2018-03-15").group, 2, "2")
      t.deepEqual(priority("IMPORTANT"), { group: 3, daysLeft: 3650 }, "{ group: 3, daysLeft: 3650 }")
      t.deepEqual(priority("no big deal").group, 4, "4")
      
;t.end()
    })

    __test("removeMarkers", t => {
      
      t.deepEqual(removeMarkers("Task DUE 2018-1-22 DUE 2018-1-11 SOON NOW"), "Task", "\"Task\"")
      t.deepEqual(removeMarkers("Task DUE 2018-1-22 DUE 2018-1-11 SOON NOW EVERY SUNDAY to do"), "Task      to do", "\"Task      to do\"")
      
;t.end()
    })

    __test("removeStarters", t => {
      
      t.deepEqual(removeStarters("### Header 1"), "Header 1", "\"Header 1\"")
      t.deepEqual(removeStarters("Not a header ###"), "Not a header ###", "\"Not a header ###\"")
      t.deepEqual(removeStarters("\t\t  * Indented"), "Indented", "\"Indented\"")
      
;t.end()
    })